src/compile/codegen.rs

```diff
@@ -4,104 +4,156 @@
-/// successful, otherwise it returns a `CompileError`.
-pub fn mavm_codegen(
-    funcs: BTreeMap<StringId, TypeCheckedFunc>,
-    string_table: &StringTable,
-    global_vars: &[GlobalVarDecl],
-    file_info_chart: &mut BTreeMap<u64, FileInfo>,
-    error_system: &mut ErrorSystem,
+/// Represents a slot number in a locals tuple
+pub type FrameSize = u32;
+pub type SlotNum = u32;
+
+/// Represents the code generation process for a function.
+struct Codegen<'a> {
+    /// The code being generated.
+    code: &'a mut Vec<Instruction>,
+    /// A source of unique labels.
+    label_gen: &'a mut LabelGenerator,
+    /// Maps `StringId`s to parsed names
+    string_table: &'a StringTable,
+    /// Associates each imported function with a unique label
+    func_labels: &'a HashMap<StringId, Label>,
+    /// List of globals this func has access to.
+    globals: &'a HashMap<StringId, GlobalVar>,
+    /// Whether to elide debug-only constructs like assert().
     release_build: bool,
-) -> Result<Vec<Instruction>, CompileError> {
-    let mut global_var_map = HashMap::new();
-    for (idx, gv) in global_vars.iter().enumerate() {
-        global_var_map.insert(gv.name_id, idx);
+
+    scopes: Vec<Scope>,
+    next_slot: SlotNum,
+}
+
+#[derive(Clone, Debug, Default)]
+struct Scope {
+    /// A variable's current slot
+    locals: BTreeMap<StringId, SlotNum>,
+    /// A variable's slot just before its first shadow
+    shadows: HashMap<StringId, SlotNum>,
+}
+
+impl Codegen<'_> {
+    fn next_slot(&mut self) -> SlotNum {
+        let next = self.next_slot;
+        self.next_slot += 1;
+        next
     }
 
-    let mut label_gen = LabelGenerator::new();
-    let mut funcs_code = BTreeMap::new();
-    for (id, func) in funcs {
-        let (lg, function_code) = mavm_codegen_func(
-            func,
-            label_gen,
-            string_table,
-            &global_var_map,
-            file_info_chart,
-            error_system,
-            release_build,
-        )?;
-        label_gen = lg;
-        funcs_code.insert(id, function_code);
+    /// Create a new scope, inheriting the locals of the one prior
+    fn open_scope(&mut self) {
+        let mut scope = match self.scopes.last() {
+            Some(scope) => scope.clone(),
+            None => Scope::default(),
+        };
+        scope.shadows = HashMap::new();
+        self.scopes.push(scope);
     }
-    let mut code = Vec::new();
-    for (_id, mut func) in funcs_code {
-        code.append(&mut func)
+
+    fn set_local(&mut self, local: StringId, slot: SlotNum) {
+        let len = self.scopes.len();
+        self.scopes[len - 1].locals.insert(local, slot);
+    }
+
+    fn shadow(&mut self, local: StringId, slot: SlotNum) {
+        let len = self.scopes.len();
+
+        // for the first time shadowing, we save the old value if it exists
+        if let Some(old) = self.scopes[len - 1].locals.get(&local) {
```

Comment:
We should consider using .get() or checking the len > 0 to avoid the possibility of a panic.

Reply:
This is outdated, I recommend pulling the latest version. We should still panic because not having a scope means something is very wrong.